File: go/day1.go
package main

import (
	"fmt"
	"time"
)

var spelledDigits = map[string]int{
	"one":   1,
	"two":   2,
	"three": 3,
	"four":  4,
	"five":  5,
	"six":   6,
	"seven": 7,
	"eight": 8,
	"nine":  9,
}

func findDigit(line string, start int, end int, step int, isSpelledDigitAllowed bool) int {
	i := start
	for (step > 0 && i < end) || (step < 0 && i > end) {
		char := line[i]
		isDigit := ('0' <= char) && (char <= '9')
		if isDigit {
			return int(char - '0')
		}

		if isSpelledDigitAllowed {
			for word, value := range spelledDigits {
				isSpelledDigit := (i+len(word) <= len(line)) && (line[i:i+len(word)] == word)
				if isSpelledDigit {
					return value
				}
			}
		}

		i += step
	}

	return 0
}

func firstDigit(line string, isSpelledDigitAllowed bool) int {
	return findDigit(line, 0, len(line), 1, isSpelledDigitAllowed)
}

func lastDigit(line string, isSpelledDigitAllowed bool) int {
	return findDigit(line, len(line)-1, -1, -1, isSpelledDigitAllowed)
}

func calculateResult(input []string, isSpelledDigitAllowed bool) int {
	res := 0
	for _, line := range input {
		value := 10*firstDigit(line, isSpelledDigitAllowed) + lastDigit(line, isSpelledDigitAllowed)
		res += value
	}
	return res
}

func Part1(input []string) int {
	return calculateResult(input, false)
}

func Part2(input []string) int {
	return calculateResult(input, true)
}

func main() {
	fileName := "input.txt"
	input := ReadFile(fileName)
	fmt.Println(Part1(input))
	fmt.Println(Part2(input))
}

func detail() {
	fileName := "input.txt"
	input := ReadFile(fileName)

	start1 := time.Now()
	fmt.Println("Answer 1 : ", Part1(input))
	fmt.Println(time.Since(start1))

	start2 := time.Now()
	fmt.Println("Answer 2 : ", Part2(input))
	fmt.Println(time.Since(start2))
}

File: python/day1.py
import re
import regex as re2

INPUT_FILE_PATH = "input.txt"
DIGIT_WORDS = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
RE_P1 = r"\d"
RE_P2 = r"\d|" + "|".join(DIGIT_WORDS)

def main(input_file):
    # Part 1
    print(sum([get_calibration_value_p1(line) \
            for line in parse_input_file(input_file)]))
    # Part 2
    print(sum([get_calibration_value_p2(line) \
            for line in parse_input_file(input_file)]))

def get_calibration_value_p1(line):
    matches = re.findall(RE_P1, line)
    first_digit = matches[0]
    last_digit = matches[-1]
    return int(first_digit + last_digit)

def get_calibration_value_p2(line):
    def get_digit_from_digit_or_word_digit(target):
        # returns (index + 1) if target in DIGIT_WORDS, else returns target
        return str(DIGIT_WORDS.index(target) + 1) if target in DIGIT_WORDS else target

    matches = re2.findall(RE_P2, line, overlapped=True)
    first_digit = matches[0]
    last_digit = matches[-1]

    first_digit = get_digit_from_digit_or_word_digit(first_digit)
    last_digit = get_digit_from_digit_or_word_digit(last_digit)

    return int(first_digit + last_digit)

def parse_input_file(input_file):
    with open(input_file, "r") as f:
        for line in f.read().split("\n"):
            line = line.strip()
            if line:
                yield line

if __name__ == "__main__":
    """
    Usage: python3 main.py [file...]
    """
    import sys
    if len(sys.argv) > 1:
        for input_file in sys.argv[1:]:
            main(input_file)
    else:
        main(INPUT_FILE_PATH)
